src/topology/continuous_function/zero_at_infty.lean

```diff
@@ -0,0 +1,318 @@
+
+instance [has_zero β] {R : Type*} [has_zero R] [smul_with_zero R β]
+  [has_continuous_const_smul R β] : smul_with_zero R (α →C₀ β) :=
+function.injective.smul_with_zero ⟨_, coe_zero⟩ fun_like.coe_injective coe_smul
+
+instance [has_zero β] {R : Type*} [monoid_with_zero R] [mul_action_with_zero R β]
+  [has_continuous_const_smul R β] : mul_action_with_zero R (α →C₀ β) :=
+function.injective.mul_action_with_zero ⟨_, coe_zero⟩ fun_like.coe_injective coe_smul
+
+instance [add_comm_monoid β] [has_continuous_add β] {R : Type*} [comm_semiring R] [module R β]
+  [has_continuous_const_smul R β] : module R (α →C₀ β) :=
+function.injective.module R ⟨_, coe_zero, coe_add⟩ fun_like.coe_injective coe_smul
+
+instance [non_unital_semiring β] [has_continuous_add β] [has_continuous_mul β] :
+  non_unital_semiring (α →C₀ β) :=
+fun_like.coe_injective.non_unital_semiring _ coe_zero coe_add coe_mul (λ _ _, rfl)
+
+end algebraic_structure
+
+section metric
+
+open metric set
+
+variables [metric_space β] [has_zero β] [zero_at_infty_continuous_map_class F α β]
+
+protected lemma bounded (f : F) : ∃ C, ∀ x y : α, dist ((f : α → β) x) (f y) ≤ C :=
+begin
+  obtain ⟨K : set α, hK₁, hK₂⟩ := mem_cocompact.mp (tendsto_def.mp (zero_at_infty (f : F)) _
+    (closed_ball_mem_nhds (0 : β) zero_lt_one)),
+  obtain ⟨C, hC⟩ := (hK₁.image (map_continuous f)).bounded.subset_ball (0 : β),
+  refine ⟨max C 1 + max C 1, (λ x y, _)⟩,
+  have : ∀ x, f x ∈ closed_ball (0 : β) (max C 1),
+  { intro x,
+    by_cases hx : x ∈ K,
+    { exact (mem_closed_ball.mp $ hC ⟨x, hx, rfl⟩).trans (le_max_left _ _) },
+    { exact (mem_closed_ball.mp $ mem_preimage.mp (hK₂ hx)).trans (le_max_right _ _) } },
+  exact (dist_triangle (f x) 0 (f y)).trans
+    (add_le_add (mem_closed_ball.mp $ this x) (mem_closed_ball'.mp $ this y)),
+end
+
+lemma bounded_range (f : α →C₀ β) : bounded (range f) :=
+bounded_range_iff.2 f.bounded
+
+lemma bounded_image (f : α →C₀ β) (s : set α) : bounded (f '' s) :=
+f.bounded_range.mono $ image_subset_range _ _
+
+@[priority 100]
+instance : bounded_continuous_map_class F α β :=
+{ coe := λ f, f,
+  coe_injective' := λ f g h, fun_like.coe_fn_eq.mp h,
+  map_continuous := λ f, map_continuous f,
+  map_bounded := λ f, zero_at_infty_continuous_map.bounded f }
+
+/-- Construct a bounded continuous function from a continuous function vanshing at infinity. -/
+@[simps]
+def to_bcf (f : α →C₀ β) : α →ᵇ β :=
+⟨f, map_bounded f⟩
+
+section
+variables (α) (β)
+lemma to_bounded_continuous_function_injective :
+  function.injective (to_bcf : (α →C₀ β) → α →ᵇ β) :=
+λ f g h, by { ext, simpa only using fun_like.congr_fun h x, }
+end
+
+variables {C : ℝ} {f g : α →C₀ β}
+
+/-- The type of continuous functions vanishing at infinity, with the uniform distance induced by the
+inclusion `zero_at_infinity_continuous_map.to_bcf`, is a metric space. -/
+noncomputable instance : metric_space (α →C₀ β) :=
+metric_space.induced _ (to_bounded_continuous_function_injective α β) (by apply_instance)
```

Comment:
Can you put a topology on `α →C₀ β` (for`β` just a topological space) by restriction of the compact-open topology, and then use the special metric space constructor which builds in compatibility with a pre-existing topology?

Reply:
This is definitely the thing to do, but it seems rather nontrivial. The reason is that `α →ᵇ β` just gets a metric structure and it is never shown (I don't think, please double-check) that the topological structure induced by the metric structure coincides with the topological structure induced by the inclusion map into `C(α, β)` with the compact-open topology. So, I think this should be done for bounded continuous functions first, and then come and do it for `C₀`.

Question: does it make sense to just define an `emetric` structure on `C(α, β)` when `β` is a metric space? This would agree with the compact open topology, right? Then we get could just redefine the metric structure on `α →ᵇ β` as the one induced by the map into `C(α, β)`.

Reply:
If we don't tackle this in this PR, I'll add a TODO to this effect.

Reply:
ah, nevermind. This is stupid because convergence in `C(α, β)` is local uniform convergence, whereas convergence in `α →ᵇ β` is uniform convergence, so these don't coincide, so this really is specific to `C₀(α, β)`.

Reply:
After more thought, the topology induced by the metric on `C₀(α, β)` does not coincide with the compact-open topology via the inclusion into `C(α, β)`. As a simple example, consider `C₀(ℕ, ℕ)` where `ℕ` has the discrete metric. Then the sequence `f : ℕ → C₀(ℕ, ℕ)` defined by `f n x = if n = x then 1 else 0` converges to `0 : C₀(ℕ, ℕ)` when equipped with the compact-open topology (because it is eventually zero on every compact, equivalently finite, set), but this sequence lies in the unit sphere centered at `0 : C₀(ℕ,ℕ)` and therefore doesn't converge in the topology induced by the metric.